File: BinaryTreePreorderTraversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def preorderTraversal(self, root):
        stack, res = [], []
        p = root
        if p != None:
            stack.append(p)
        while stack:
            p = stack.pop()
            res.append(p.val)
            if p.right != None:
                stack.append(p.right)
            if p.left != None:
                stack.append(p.left)
        return res

File: CPP/Second/binary_tree_preorder_traversal.cc
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        stack<TreeNode *> s;
        vector<int> res;
        TreeNode *p;
        p = root;
        if (p != nullptr)
            s.push(p);
            
        while(!s.empty()) {
            p = s.top();
            s.pop();
            res.push_back(p->val);
            
            if (p->right)
                s.push(p->right);
            if (p->left)
                s.push(p->left);
        }
        return res;
    }
};

int main(int argc, char *argv[])
{

    return 0;
}
